<!DOCTYPE html>
<html>

<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>run_treemix.md</title>
  <link rel="stylesheet" href="https://stackedit.io/style.css" />
</head>

<body class="stackedit">
  <div class="stackedit__html"><h2 id="scalepopgen-treemix">scalepopgen: TreeMix</h2>
<p>This sub-workflow can be invoked using the argument <code>treemix = true</code>. It consists of the following steps:</p>
<p><strong>1.</strong> converting vcf to treemix input file format using python script (vcf_to_treemix.py)</p>
<p><strong>2.</strong> merging converted treemix files (separated by chromosomes) into one</p>
<p><strong>3.</strong> generating a file with samples and populations based on sample map provided by the user</p>
<p><strong>4.</strong> running the treemix analysis without any migration edge</p>
<p><strong>5.</strong> running the treemix analysis with 100 bootstraps</p>
<p><strong>6.</strong> generating consensus tree out of the 100 bootstrapped trees generated in the previous step</p>
<p><strong>7.</strong> running iterations of treemix algorithm for each migration edge (In case of 10 iterations and 10 migration edges, the program will run 10 iterations for migration edges starting from 1 to 10.)</p>
<p><strong>8.</strong> identify optimal migration edges using the procedure as implemented in <a href="https://academic.oup.com/biomethods/article/6/1/bpab017/6371180">OptM</a> package</p>
<blockquote>
<p><strong>Note:</strong> If your input files are in Plink format, it will convert them into VCF before the first step.</p>
</blockquote>
<h2 id="description-of-the-parameters">Description of the parameters:</h2>
<p><code>n_bootstrap</code>: number of bootstrapping to be performed (default: 100); if the number is set to 0, bootstraping will not be performed;<br>
<code>upper_limit</code>: the number of bootstrapping to be performed (default: 100);<br>
<code>starting_m_value</code>:  the starting value of migration edges to be added (default: 1); if the number is set to 0, it will skip the steps of adding migration edges (steps 7 and 8);<br>
<code>ending_m_value</code>: the ending value of migration edges to be added (default: 10);<br>
<code>n_iter</code>: the number of iterations to be performed for each migration edge (default: 10);<br>
<code>k_snps</code>: the block of SNPs to be grouped together to account for the LD (default: 500);</p>
<h2 id="description-of-the-output-files-generated-by-this-sub-workflow">Description of the output files generated by this sub-workflow:</h2>
<p>If the pipeline has completed successfully, all these output files will be stored in <strong>${output directory}/treemix/</strong>:<br>
<strong>*_treemix_input.gz</strong> : merged input file used for the treemix analysis (step 2)<br>
<strong>*_default_mig0_*</strong> : output files from the treemix analysis described in the step 4 above.<br>
<strong>*<em>bootstrap_iter</em>${randit}_</strong>* : output files from the treemix analysis described in step 5 above; “randit” refers to the random seed set for each bootstrapping<br>
<strong>treefile</strong> and <strong>treeout</strong> : consensus tree info produced after bootstrapping (step 6)<br>
<strong>*_default_${it}_m${mig}_*</strong> : output files from the treemix analysis described in the step 7; the variable “it” refers to the number of iteration and “mig” refers to the number of migration edges considered.<br>
<strong>*_optm_out.pdf</strong> : output files from the program Optm as described in the step 8 above</p>
<blockquote>
<p><strong>Note:</strong> The output folder also contains the pdf files of output tree and residual matrix of each treemix run.</p>
</blockquote>
<h2 id="validation-of-the-sub-workflow">Validation of the sub-workflow:</h2>
<h3 id="required-input-data-files">1. Required input data files</h3>
<p>For workflow validation, we have downloaded publicly available samples with whole genome sequences from NCBI database. We included domestic goats (<em>Capra hircus</em>) represented by various breeds from Morocco and Switzerland (geo_map). In addition to them, we also included some wild <em>Capra</em> members that are: Alpine ibex (<em>C. ibex</em>), Iberian ibex (<em>C. pyrenaica</em>) and Siberian ibex (<em>C. sibirica</em>). Since we need an outgroup when performing some analyses, we also added Urial sheep (<em>Ovis vignei</em>). We will use variants from chromosome 28 and 29 of, all together, 39 animals.</p>
<p>The input data should be in the <strong>VCF</strong> or <strong>PLINK binary</strong> format files.</p>
<p>All VCF files need to be splitted by the chromosomes and indexed with tabix. You will have to prepare csv list of those files. Each row is corresponding to one chromosome and each row has three different information separated by the comma. Like in example below, the first information in each row is chromosome name, next is path/to/the/file.vcf.gz and the last is path/to/the/file.vcf.gz.tbi.</p>
<pre><code>Chr28,/data/trial/NC_030835.1.vcf.gz,/data/trial/NC_030835.1.vcf.gz.tbi
Chr29,/data/trial/NC_030836.1.vcf.gz,/data/trial/NC_030836.1.vcf.gz.tbi
</code></pre>
<p>In addition to the VCF input format, it is also necessary to prepare a sample map file of individuals and populations. Sample map has two tab-delimited columns: in the first column are individual IDs and in the second are population IDs as demonstrated on the example below.</p>
<pre><code>AfrGoat10 AfricaGoat
AfrGoat1 AfricaGoat
AfrGoat2 AfricaGoat
AfrGoat3 AfricaGoat
AfrGoat4 AfricaGoat
AfrGoat5 AfricaGoat
AfrGoat6 AfricaGoat
AfrGoat7 AfricaGoat
AfrGoat8 AfricaGoat
AfrGoat9 AfricaGoat
AlpIbex1 AlpineIbex
AlpIbex2 AlpineIbex
AlpIbex3 AlpineIbex
AlpIbex4 AlpineIbex
AlpIbex5 AlpineIbex
AlpIbex6 AlpineIbex
AlpIbex7 AlpineIbex
AlpIbex8 AlpineIbex
AlpIbex9 AlpineIbex
IbrIbex1 IberianIbex
IbrIbex2 IberianIbex
IbrIbex3 IberianIbex
IbrIbex4 IberianIbex
SibIbex1 SiberianIbex
SibIbex2 SiberianIbex
SwiGoat10 SwissGoat
SwiGoat11 SwissGoat
SwiGoat12 SwissGoat
SwiGoat1 SwissGoat
SwiGoat2 SwissGoat
SwiGoat3 SwissGoat
SwiGoat4 SwissGoat
SwiGoat5 SwissGoat
SwiGoat6 SwissGoat
SwiGoat7 SwissGoat
SwiGoat8 SwissGoat
SwiGoat9 SwissGoat
UriShep1 Urial
UriShep2 Urial
</code></pre>
<p>For the Plink binary input, you need to specified the path to the BED/BIM/FAM files in the section of general parameters:<br>
<code>input= "path/to/the/files/*.{bed,bim,fam}"</code></p>
<h3 id="setting-the-parameters">2. Setting the parameters</h3>
<p>At the beginning of the parameter file *<strong>/parameters/scale_popgen.config</strong>, we have to specified some of the general things first.<br>
<code>input</code>: name of the .csv input file for the VCF format or names of the PLINK binary files<br>
<code>outDir</code>: the name of the output folder<br>
<code>prefix</code>: prefix of the output files<br>
<code>sample_map</code>: the name of the file with individuals and populations as addition to VCF input<br>
<code>fasta</code>: the name of the reference genome fasta file that will be used for converting from VCF to PLINK format<br>
<code>allow_extra_chrom</code>: set to true if the input contains chromosome name in the form of string<br>
<code>max_chrom</code>: maximum number of chromosomes<br>
<code>outgroup</code>: the population ID of the outgroup (optimal)<br>
<code>cm_to_bp</code>: the number of base pairs that corresponds to one centimorgan (optimal)</p>
<p>After that, there are two parts dedicated to the TreeMix analysis.</p>
<p><strong>// general parameters</strong></p>
<pre><code>input                     = "/data/trial/input.csv"
outDir                    = "${baseDir}/../TreeMix/"
prefix                    = "TreeMix"
sample_map                = "/data/trial/sample.map"
fasta                     = "/data/trial/goat_ref.fasta"
chrm_map                  = "none"
allow_extra_chrom         = true 
max_chrom                 = 2
outgroup                  = "Urial"
cm_to_bp                   = 0
</code></pre>
<p><strong>//treemix analysis parameters</strong></p>
<pre><code>treemix                   = true
n_bootstrap               = 10
upper_limit               = 600
starting_m_value          = 1
ending_m_value            = 5
n_iter                    = 10
k_snps                    = 5000
</code></pre>
<h2 id="references">References</h2>
<p>Please cite the following papers if you use this sub-workflow in your study:</p>
<p>[1] Pickrell JK, Pritchard JK (2012) Inference of Population Splits and Mixtures from Genome-Wide Allele Frequency Data. PLOS Genetics 8(11): e1002967. <a href="https://doi.org/10.1371/journal.pgen.1002967">https://doi.org/10.1371/journal.pgen.1002967</a><br>
[2] Robert R Fitak, OptM: estimating the optimal number of migration edges on population trees using Treemix, Biology Methods and Protocols, Volume 6, Issue 1, 2021, bpab017, <a href="https://doi.org/10.1093">https://doi.org/10.1093</a><br>
[3] Di Tommaso, P., Chatzou, M., Floden, E. et al. Nextflow enables reproducible computational workflows. Nat Biotechnol 35, 316â€“319 (2017). <a href="https://doi.org/10.1038/nbt.3820">https://doi.org/10.1038/nbt.3820</a><br>
[4] Upadhyay M., Pogorevc N., Medugorac I. (2023). scalepopgen: a bioinformatics workflow resources implemented in Nextflow for population genomic analyses. XXXXXXXXXXXXX</p>
<h2 id="license">License</h2>
<p>MIT</p>
</div>
</body>

</html>
