<!DOCTYPE html>
<html>

<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>run_pca_admixture.md</title>
  <link rel="stylesheet" href="https://stackedit.io/style.css" />
</head>

<body class="stackedit">
  <div class="stackedit__html"><h2 id="scalepopgen-pca-and-admixture">scalepopgen: PCA and ADMIXTURE</h2>
<p>This module combines Admixture analysis and two different Principal Component Analyses. The first PCA can be done with the function SmartPCA of the <a href="https://github.com/chrchang/eigensoft/tree/master/POPGEN">EIGENSOFT</a> software and performed with the argument: <code>run_smartpca = true</code>. The second one is done by using the function snpgdsPCA inside the R package <a href="https://code.bioconductor.org/browse/SNPRelate/RELEASE_3_17/">SNPRelate</a> and can be invoked with the argument: <code>run_gds_pca = true</code>. Analysis with the program <a href="https://dalexander.github.io/admixture/">Admixture</a> are carried out with the argument <code>admixture = true</code>.</p>
<p>The above-mentioned analyses consist of the following steps:</p>
<p><strong>1.</strong> converting vcf to bed file format using Plink<br>
<strong>2.</strong> merging converted bed files (separated by chromosomes) into one with Plink<br>
<strong>3.</strong> removing listed individuals and updating the family IDs with Plink<br>
<strong>4.</strong> applying Linkage Disequilibrium based filtering of bed files with Plink<br>
<strong>5.</strong> running both types of PCA analyses with plotting the results<br>
<strong>6.</strong> running admixture from the starting and ending k-value selected by the user<br>
<strong>7.</strong> performing block cross-validation of k-values<br>
<strong>8.</strong> selecting the optimal number of clusters (K) regarding the lowest cross validation error and plotting the results</p>
<blockquote>
<p><strong>Note:</strong> If your input files are already in Plink format, it will skip the first step.</p>
</blockquote>
<h2 id="description-of-the-parameters">Description of the parameters:</h2>
<p><code>ld_filt</code>: an option to use LD-based filtering according to parameters specified below (default: true), meaning to include or skip step 4;<br>
<code>ld_window_size</code>: a window size (default: 50) in variant count or kilobases (step 4);<br>
<code>step_size</code>: number of variants (default: 10) to shift the window at the end of each step (step 4);<br>
<code>r2_value</code>: squared correlation threshold (default: 0.1); at each step, only pairs of variants with r2 greater than the threshold are recognized (step 4);<br>
<code>structure_remove_indi</code>: the name of a file in which you specified samples you do not want to use for admixture and PCA (step 5);<br>
<code>smartpca_param</code>: the name of a parameter file that will be used for the SmartPCA (step 5);<br>
<code>pop_color_file</code>: a name of a tab-delimited text file with specified R colors for each population; the Family IDs are in the first column and R-color code in the second (step 5);<br>
<code>starting_k_value</code>: starting number of clusters (step 6);<br>
<code>ending_k_value</code>: maximal number of clusters (step 6);<br>
<code>method</code>: type of the method (default: “block”) to select the optimal number of clusters (step 7), option “-m” in program Admixture - alternative to “block” relaxation algorithm is “EM” algorithm;<br>
<code>cross_validation</code>: the number of folds (default: 10) for cross-validation (step 7);<br>
<code>termination_criteria</code>: the lowest limit of the log-likelihood change (default: 0.0001) between iterations to stop the process (step 7), option “-C” in program Admixture - based on integer or floating number this argument is interpreted as a delta or an iteration count;<br>
<code>best_kval_method</code>: the procedure (default: “global”) for choosing the optimal K value (step 8)<br>
<code>pop_labes</code>: the name of the text file; you use this option, if you want to change population IDs for the admixture graph (step 8)</p>
<h2 id="description-of-the-output-files-generated-by-this-sub-workflow">Description of the output files generated by this sub-workflow:</h2>
<p>If the pipeline has completed successfully, all these output files will be stored in <strong>${output directory}/genetic_structure/</strong>.</p>
<p>You will find the PCA results in two sub-folder <strong>./smartpca</strong> and <strong>./gds_pca</strong>.<br>
Results of SmartPCA:<br>
<strong>*_ld_filtered.snp</strong> , <strong>*_ld_filtered.ind</strong>, <strong>*_ld_filtered.eigenstratgeno</strong> : input files used for the analysis<br>
<strong>*_ld_filtered.evec</strong> : a table of a positions for each individual along  with eigenvectors in the header row<br>
<strong>*_ld_filtered.eval</strong> : a column of ordered eigenvalues corresponding to the eigenvectors<br>
Results of snpgdsPCA:<br>
<strong>*_ld_filtered.varprop</strong> : proportion of variance for each principal component<br>
<strong>*_ld_filtered.jpeg</strong> : graph of first two eigenvalues<br>
<strong>*_ld_filtered.eigenvect</strong> : a table of eigenvectors for each individual</p>
<p>You will find the admixture results in a sub-folder <strong>./admixture</strong>:<br>
<strong>*_ld_filtered.${K}.P</strong> : table of the allele frequencies inferred for each SNP in each population<br>
<strong>*_ld_filtered.${K}.Q</strong> : table of inferred individual ancestry proportions from the K ancestral populations, with one row per individual<br>
<strong>*.${K}.Q.png</strong> : the admixture plot of the optimal K value<br>
<strong>\best_k_${K}.png</strong> : the plot of CV errors for determinating optimal K value<br>
<strong>pongInput.map</strong> : the input file for the program Pong.</p>
<h2 id="validation-of-the-sub-workflow">Validation of the sub-workflow:</h2>
<h3 id="required-input-data-files">1. Required input data files</h3>
<p>For workflow validation, we have downloaded publicly available samples with whole genome sequences from NCBI database. We included domestic goats (<em>Capra hircus</em>) represented by various breeds from Morocco and Switzerland (geo_map). In addition to them, we also included some wild <em>Capra</em> members that are: Alpine ibex (<em>C. ibex</em>), Iberian ibex (<em>C. pyrenaica</em>) and Siberian ibex (<em>C. sibirica</em>). Since we need an outgroup when performing some analyses, we also added Urial sheep (<em>Ovis vignei</em>). We will use variants from chromosome 28 and 29 of, all together, 39 animals.</p>
<p>The input data should be in the <strong>VCF</strong> or <strong>PLINK binary</strong> format files.</p>
<p>All VCF files need to be splitted by the chromosomes and indexed with tabix. You will have to prepare csv list of those files. Each row is corresponding to one chromosome and each row has three different information separated by the comma. Like in example below, the first information in each row is chromosome name, next is path/to/the/file.vcf.gz and the last is path/to/the/file.vcf.gz.tbi.</p>
<pre><code>Chr28,/data/trial/NC_030835.1.vcf.gz,/data/trial/NC_030835.1.vcf.gz.tbi
Chr29,/data/trial/NC_030836.1.vcf.gz,/data/trial/NC_030836.1.vcf.gz.tbi
</code></pre>
<p>In addition to the VCF input format, it is also necessary to prepare a sample map file of individuals and populations. Sample map has two tab-delimited columns: in the first column are individual IDs and in the second are population IDs as demonstrated on the example below.</p>
<pre><code>AfrGoat10 AfricaGoat
AfrGoat1 AfricaGoat
AfrGoat2 AfricaGoat
AfrGoat3 AfricaGoat
AfrGoat4 AfricaGoat
AfrGoat5 AfricaGoat
AfrGoat6 AfricaGoat
AfrGoat7 AfricaGoat
AfrGoat8 AfricaGoat
AfrGoat9 AfricaGoat
AlpIbex1 AlpineIbex
AlpIbex2 AlpineIbex
AlpIbex3 AlpineIbex
AlpIbex4 AlpineIbex
AlpIbex5 AlpineIbex
AlpIbex6 AlpineIbex
AlpIbex7 AlpineIbex
AlpIbex8 AlpineIbex
AlpIbex9 AlpineIbex
IbrIbex1 IberianIbex
IbrIbex2 IberianIbex
IbrIbex3 IberianIbex
IbrIbex4 IberianIbex
SibIbex1 SiberianIbex
SibIbex2 SiberianIbex
SwiGoat10 SwissGoat
SwiGoat11 SwissGoat
SwiGoat12 SwissGoat
SwiGoat1 SwissGoat
SwiGoat2 SwissGoat
SwiGoat3 SwissGoat
SwiGoat4 SwissGoat
SwiGoat5 SwissGoat
SwiGoat6 SwissGoat
SwiGoat7 SwissGoat
SwiGoat8 SwissGoat
SwiGoat9 SwissGoat
UriShep1 Urial
UriShep2 Urial
</code></pre>
<p>For the Plink binary input, you need to specified the path to the BED/BIM/FAM files in the section of general parameters:<br>
<code>input= "path/to/the/files/*.{bed,bim,fam}"</code></p>
<h3 id="optional-input-data-files">2. Optional input data files</h3>
<p>This module allows you to remove samples, like the outgroup, that you want to exclude in given analyses (<code>structure_remove_indi</code>). You have to prepare a space/tab-delimited text file with family IDs in the first column and within-family IDs in the second column as stated by <a href="https://www.cog-genomics.org/plink/2.0/filter#sample">PLINK</a> (option --remove):</p>
<pre><code>Urial UriShep1
Urial UriShep2
</code></pre>
<p>If you want to provide your file with optional parameters for the program SmartPCA (<code>smartpca_param</code>). You can make one according to the <a href="https://github.com/chrchang/eigensoft/tree/master/POPGEN">instructions of the EIGENSOFT software</a>.</p>
<p>As well, you can specify the desired colors by population for the final PCA graphs (<code>pop_color_file</code>). You have to make a tab-delimited text file, where the population names are in the first column and specified R color codes in the second:</p>
<p>Additional you can also change population names (<code>pop_labels</code>), which will be considered only for plotting the results. You need to prepare a text file…to be continued…</p>
<h3 id="setting-the-parameters">3. Setting the parameters</h3>
<p>At the beginning of the parameter file *<strong>/parameters/scale_popgen.config</strong>, we have to specified some of the general things first.<br>
<code>input</code>: name of the .csv input file for the VCF format or names of the PLINK binary files<br>
<code>outDir</code>: the name of the output folder<br>
<code>prefix</code>: prefix of the output files<br>
<code>sample_map</code>: the name of the file with individuals and populations as addition to VCF input<br>
<code>fasta</code>: the name of the reference genome fasta file that will be used for converting from VCF to PLINK format<br>
<code>allow_extra_chrom</code>: set to true if the input contains chromosome name in the form of string<br>
<code>max_chrom</code>: maximum number of chromosomes<br>
<code>outgroup</code>: the population ID of the outgroup (optimal)<br>
<code>cm_to_bp</code>: the number of base pairs that corresponds to one centimorgan (optimal)</p>
<p>After that, there are two parts dedicated to the PCA and ADMIXTURE analyses.</p>
<p><strong>// general parameters</strong></p>
<pre><code>input                     = "/data/trial/input.csv"
outDir                    = "${baseDir}/../PCA_Admix/"
prefix                    = "PCA_Admix"
sample_map                = "/data/trial/sample.map"
fasta                     = "/data/trial/goat_ref.fasta"
chrm_map                  = "none"
allow_extra_chrom         = true 
max_chrom                 = 2
outgroup                  = "Urial"
cm_to_bp                   = 0
</code></pre>
<p><strong>//ld filter for PCA and ADMIXTURE analysis</strong></p>
<pre><code>run_smartpca             = true
run_gds_pca              = true
ld_filt                  = true
ld_window_size           = 50
step_size                = 10
r2_value                 = 0.01

structure_remove_indi    = "remov_outgroup.txt"
smartpca_param           = "none"
pop_color_file           = "none"
</code></pre>
<p><strong>//admixture analysis parameters</strong></p>
<pre><code>admixture                 = true
starting_k_value          = 2
ending_k_value            = 10
method                    = "block"
cross_validation          = 5
termination_criteria      = 0.0001
best_kval_method          = "global"
pop_labels                = "none"
</code></pre>
<h2 id="references">References</h2>
<p>Please cite the following papers if you use this sub-workflow in your study:</p>
<p>[1] Alexander, D. H., Novembre, J., &amp; Lange, K. (2009). Fast model-based estimation of ancestry in unrelated individuals. Genome research, 19(9), 1655â€“1664. <a href="https://doi.org/10.1101/gr.094052.109">https://doi.org/10.1101/gr.094052.109</a></p>
<p>[2] Patterson, N., Price, A. L., &amp; Reich, D. (2006). Population structure and eigenanalysis. PLoS genetics, 2(12), e190. <a href="https://doi.org/10.1371/journal.pgen.0020190">https://doi.org/10.1371/journal.pgen.0020190</a></p>
<p>[3] Price, A. L., Patterson, N. J., Plenge, R. M., Weinblatt, M. E., Shadick, N. A., &amp; Reich, D. (2006). Principal components analysis corrects for stratification in genome-wide association studies. Nature genetics, 38(8), 904â€“909. <a href="https://doi.org/10.1038/ng1847">https://doi.org/10.1038/ng1847</a></p>
<p>[4] Zheng, X., Levine, D., Shen, J., Gogarten, S. M., Laurie, C., &amp; Weir, B. S. (2012). A high-performance computing toolset for relatedness and principal component analysis of SNP data. Bioinformatics (Oxford, England), 28(24), 3326â€“3328. <a href="https://doi.org/10.1093/bioinformatics/bts606">https://doi.org/10.1093/bioinformatics/bts606</a></p>
<h2 id="license">License</h2>
<p>MIT</p>
</div>
</body>

</html>
